src/CoreWCF.Primitives/src/CoreWCF/Channels/ChannelDemuxer.cs

```diff
@@ -0,0 +1,3848 @@
+﻿using CoreWCF.Configuration;
+using CoreWCF.Dispatcher;
+using CoreWCF.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CoreWCF.Channels
+{
+    class ChannelDemuxer
+    {
+        public readonly static TimeSpan UseDefaultReceiveTimeout = TimeSpan.MinValue;
+
+        TypedChannelDemuxer inputDemuxer;
+        TypedChannelDemuxer replyDemuxer;
+        Dictionary<Type, TypedChannelDemuxer> typeDemuxers;
+        TimeSpan peekTimeout;
+        int maxPendingSessions;
+
+        public ChannelDemuxer()
+        {
+            this.peekTimeout = ChannelDemuxer.UseDefaultReceiveTimeout; //use the default receive timeout (original behavior)
+            this.maxPendingSessions = 10;
+            this.typeDemuxers = new Dictionary<Type, TypedChannelDemuxer>();
+        }
+
+        public TimeSpan PeekTimeout
+        {
+            get
+            {
+                return this.peekTimeout;
+            }
+            set
+            {
+                this.peekTimeout = value;
+            }
+        }
+
+        public int MaxPendingSessions
+        {
+            get
+            {
+                return this.maxPendingSessions;
+            }
+            set
+            {
+                this.maxPendingSessions = value;
+            }
+        }
+
+       internal IServiceDispatcher CreaterServiceDispatcher<TChannel>(IServiceDispatcher innerDispatcher, ChannelDemuxerFilter filter)
+        {
+            return GetTypedServiceDispatcher<TChannel>().AddDispatcher(innerDispatcher, filter);
+        }
+
+        internal IServiceDispatcher CreaterServiceDispatcher<TChannel>(IServiceDispatcher innerDispatcher)
+        {
+            return GetTypedServiceDispatcher<TChannel>().AddDispatcher(innerDispatcher, new ChannelDemuxerFilter(new MatchAllMessageFilter(), 0));
+        }
+
+        internal void RemoveServiceDispatcher<TChannel>(MessageFilter filter)
+        {
+             GetTypedServiceDispatcher<TChannel>().RemoveDispatcher(filter);
+        }
+        internal TypedChannelDemuxer GetTypedServiceDispatcher<TChannel>()
+        {
+            TypedChannelDemuxer typeDemuxer = null;
+            bool createdDemuxer = false;
+
+            //if (typeof(TChannel) == typeof(IInputChannel))
```

Comment:
Can you open an issue for needing to support channel demuxer for channel shapes other than IReplyChannel

Reply:
https://github.com/CoreWCF/CoreWCF/issues/251